### PR TITLE
feat(firestore-bigquery-export): Add Gemini agent to gen-schema-view

### PR DESCRIPTION
**Pull request by cabljac**
This PR adds the option to use a Gemini Agent to generate the schema files, fixes some tests, and corrects a parameter description.

### CODE REVIEW

File: firestore-bigquery-export/scripts/gen-schema-view/src/schema/index.ts [hunk 1]
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+// TODO: collection name vs table name prefix split them up, they're different.
```
Comment (cabljac) [inline firestore-bigquery-export/scripts/gen-schema-view/src/schema/index.ts:17]:
```suggestion
```

**Comment by Gustolandia** [comment]
These are some suggestions that I think you can go through before merging this PR:

### **1. Logging for a Production Environment**
**Areas Missing Logs:**
- **`src/schema/genkit.ts`:**
  - **Logging Schema Validation Failures:**
    - Current validation errors in `writeSchemaTool` only return a string, but there's no log statement for debugging in production environments.
      - **Suggestion:** Add a `console.error` or a structured log for schema validation failures to capture the details of the `error.message` and the invalid `content`.
        ```typescript
        } catch (error) {
            console.error("Schema validation failed:", { error: error.message, content });
            return `Error creating schema: ${error.message}`;
        }
        ```
  - **Logging File Write Errors:**
    - If `fs.writeFile` fails, there’s no mechanism to log why it failed. A `try-catch` should be added to log these errors.
      ```typescript
      try {
          await fs.writeFile(filePath, content);
      } catch (writeError) {
          console.error("Error writing schema file:", { filePath, error: writeError.message });
          throw writeError;
      }
      ```

- **Agent Execution:**
  - **Missing Logs:** In the `runAgent` function, there's no logging to track:
    1. Whether the schema was successfully saved.
    2. If the AI model encountered any issues during analysis or generation.
    3. The processing time for schema generation.
    - **Suggestion:** Add logs at key stages.
      ```typescript
      console.log("Starting schema generation with Gemini agent...");
      const startTime = Date.now();
      const response = await ai.chat(schemaAgent);
      console.log("Agent response received. Duration:", Date.now() - startTime, "ms");
      if (response.error) {
          console.error("Error in schema generation:", response.error);
      } else {
          console.log("Schema generation completed successfully.");
      }
      ```

---

### **2. Testing**
**Gaps in Testing:**
- **Edge Case Scenarios:**
  - **Deeply Nested Documents:**
    - No explicit test cases for collections with deeply nested fields, such as maps within arrays or arrays within maps.
    - **Suggestion:** Add a test in `e2e.test.ts` with a deeply nested document like:
      ```typescript
      const sampleData = [
          {
              user: {
                  profile: {
                      name: "Alice",
                      age: 25,
                  },
                  settings: {
                      notifications: {
                          email: true,
                          push: false,
                      },
                  },
              },
          },
      ];
      ```

  - **Large Collections:**
    - There are no tests for large collections that exceed the `agent-sample-size`. A test case should simulate the agent gracefully sampling the first N documents and stopping without errors.
      ```typescript
      test("should handle collections with more than 100 documents", async () => {
          const largeSampleData = Array.from({ length: 200 }, (_, i) => ({ id: i, value: `doc${i}` }));
          const response = await runAgent(apiKey, "./schemas", "large_collection", "test_table", largeSampleData);
          expect(response).toBeDefined();
          expect(response.success).toBeTruthy();
      });
      ```

- **Invalid Data Handling:**
  - Test scenarios with inconsistent or invalid Firestore documents are missing.
    - **Suggestion:** Add a test case where documents have inconsistent field types:
      ```typescript
      const sampleData = [
          { field1: "string" },
          { field1: 123 }, // inconsistent type
      ];
      ```

- **Error Cases in AI Model Execution:**
  - No tests simulate failures in the Gemini AI API.
    - **Suggestion:** Mock the API to return errors, such as rate limits or timeouts, and ensure these are handled gracefully.

- **File Overwrite Edge Case:**
  - The agent currently checks for existing files but doesn't test scenarios where users want to overwrite.
    - **Suggestion:** Add a test to validate overwrite handling logic:
      ```typescript
      test("should not overwrite existing schema files", async () => {
          // Mock an existing file scenario
          jest.spyOn(fs, "access").mockResolvedValueOnce(true);
          const response = await runAgent(apiKey, "./schemas", "test_collection", "test_table", []);
          expect(response).toContain("Error: Schema file already exists");
      });
      ```

---

### **3. Refactoring**
**Areas Needing Refactor:**
- **`src/schema/genkit.ts`:**
  - **Monolithic Functionality in `runAgent`:**
    - The `runAgent` function handles too many responsibilities: initializing AI tools, validating schemas, generating schemas, and saving files.
    - **Suggestion:** Break it into smaller, more focused functions, such as:
      1. `initializeAI(apiKey): AI`
      2. `validateSchema(schema: string): boolean`
      3. `generateSchema(agent, sampleData): string`
      4. `saveSchema(filePath, content): void`
      ```typescript
      const runAgent = async (apiKey, schemaDirectory, collectionName, tablePrefix, sampleData) => {
          const ai = initializeAI(apiKey);
          const schemaContent = generateSchema(ai, sampleData);
          const isValid = validateSchema(schemaContent);
          if (!isValid) {
              console.error("Invalid schema generated.");
              return;
          }
          saveSchema(path.join(schemaDirectory, `${tablePrefix}.json`), schemaContent);
      };
      ```

---

### **4. Documentation**
**Missing Examples:**
- **Custom Schema Adjustments:**
  - Users may want to modify the schema post-generation, such as overriding field names or adding metadata.
    - **Suggestion:** Include an example of modifying the generated schema:
      ```bash
      # Example: Adjusting generated schema
      npx @firebaseextensions/fs-bq-schema-views \
          --use-gemini-agent \
          --collection-path=my_collection \
          --schema-dir=./schemas
      # After generation, edit ./schemas/my_table.json to add custom metadata or adjust field names.
      ```

- **Error Debugging:**
  - Provide examples of interpreting and resolving common errors, such as API key issues or validation failures.
    - **Suggestion:** Add a “Troubleshooting” section:
      ```
      ### Troubleshooting
      - **Error: Schema validation failed**: Ensure the sample documents reflect the structure of the collection.
      - **Error: API key invalid**: Verify the `--google-ai-key` value is correct and authorized.
      ```

- **Advanced Use Cases:**
  - Missing examples for advanced scenarios like:
    1. Collections with mixed field types.
    2. Using the agent to update schemas incrementally.
    - **Suggestion:** Add an example:
      ```bash
      npx @firebaseextensions/fs-bq-schema-views \
          --use-gemini-agent \
          --collection-path=my_collection \
          --schema-dir=./schemas \
          --agent-sample-size=20
      ```

**Comment by cabljac** [comment]
>   * **Deeply Nested Documents:**
>     
>     * No explicit test cases for collections with deeply nested fields, such as maps within arrays or arrays within maps.
>     * **Suggestion:** Add a test in `e2e.test.ts` with a deeply nested document like:
>       ```ts
>       const sampleData = [
>           {
>               user: {
>                   profile: {
>                       name: "Alice",
>                       age: 25,
>                   },
>                   settings: {
>                       notifications: {
>                           email: true,
>                           push: false,
>                       },
>                   },
>               },
>           },
>       ];
>       ```
>   * **Large Collections:**
>     
>     * There are no tests for large collections that exceed the `agent-sample-size`. A test case should simulate the agent gracefully sampling the first N documents and stopping without errors.
>       ```ts
>       test("should handle collections with more than 100 documents", async () => {
>           const largeSampleData = Array.from({ length: 200 }, (_, i) => ({ id: i, value: `doc${i}` }));
>           const response = await runAgent(apiKey, "./schemas", "large_collection", "test_table", largeSampleData);
>           expect(response).toBeDefined();
>           expect(response.success).toBeTruthy();
>       });
>       ```


I think i'll leave out these, they're fair points but not relevant to this PR in particular.

Actually the "Large Collections" comment isn't correct; we're taking just a sample

**Comment by Gustolandia** [comment]
> Actually the "Large Collections" comment isn't correct; we're taking just a sample

The current tests seem to be missing a verification for whether the sampling behavior correctly handles large collections. The tests should explicitly check that only the specified number of samples (agent-sample-size) are being processed, regardless of the total size of the collection.

What I wrote should be like this instead (apologies):

```typescript
      test("should handle collections with more than 100 documents by sampling", async () => {
          const largeSampleData = Array.from({ length: 200 }, (_, i) => ({ id: i, value: `doc${i}` }));
          const agentSampleSize = 50; // Example sample size
          const response = await runAgent(apiKey, "./schemas", "large_collection", "test_table", largeSampleData);

          // Mocking the sampling logic to verify correct behavior
          const sampledData = largeSampleData.slice(0, agentSampleSize); // Expected sample
          expect(response).toBeDefined();
          expect(response.success).toBeTruthy();

          // Assert that only sampledData is used for schema generation
          expect(response.processedData).toEqual(sampledData); // Replace with actual implementation if available
      });
      ```

File: firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md [hunk 2]
```diff
@@ -314,6 +314,62 @@ Since all document data is stored in the schemaless changelog, mistakes in
 schema configuration don't affect the underlying data and can be resolved by
 re-running the schema-views script against an updated schema file.
 
+### Using Gemini to generate draft schema files
+
+Instead of manually creating schema files, you can use the built-in Gemini AI Agent to automatically analyze your Firestore collection and generate an appropriate schema. The agent will:
```
Comment (pr-Mais) [inline firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md:319]:
```suggestion
Instead of manually creating schema files, you can use the built-in Gemini Agent to automatically analyze your Firestore collection and generate an appropriate schema. The agent will:
```

File: firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md [hunk 3]
```diff
@@ -314,6 +314,62 @@ Since all document data is stored in the schemaless changelog, mistakes in
...
+1. Sample documents from your collection
+2. Analyze the data structure
+3. Generate a well-documented schema file
+4. Create the corresponding BigQuery views
+
+You can use Gemini in either interactive or non-interactive mode:
+
+```bash
+# Interactive mode
+npx @firebaseextensions/fs-bq-schema-views
```
Comment (pr-Mais) [inline firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md:330]: I think its better to include a copy-paste minimal usage here, e.g. add `--use-gemini-agent` and other required params

File: firestore-bigquery-export/scripts/gen-schema-view/src/index.ts [hunk 4]
```diff
@@ -123,40 +234,109 @@ interface CliConfig {
...
   );
-  for (const schemaName in config.schemas) {
-    await viewFactory.initializeSchemaViewResources(
-      config.datasetId,
-      config.tableNamePrefix,
-      schemaName,
-      config.schemas[schemaName]
-    );
+
+  if (config.useGemini) {
```
Comment (pr-Mais) [inline firestore-bigquery-export/scripts/gen-schema-view/src/index.ts:262]: should we check if the schema file already exist and exit?

**Review by pr-Mais — CHANGES_REQUESTED** [review]

File: firestore-bigquery-export/scripts/gen-schema-view/src/index.ts (hunk 4 above)
Comment (cabljac) [reply firestore-bigquery-export/scripts/gen-schema-view/src/index.ts:262]: yes, good catch

File: firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md (hunk 3 above)
Comment (cabljac) [reply firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md:330]: We have a non-interactive example below, can you explain what you mean? confused

File: firestore-bigquery-export/scripts/gen-schema-view/src/__tests__/e2e/e2e.legacy.test.ts [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+    /** Get the query result */
+    const result = await getQueryResult(dataset.id, tableId, schemaName);
+
+    /** Assert data */
+    expect(result.length).toBe(1);
+    expect(result[0].name).toBe("test");
+  }, 80000);
+
+  test("should generate a schame view based on a nestedMapSchema dataset and schema", async () => {
```
Comment (cabljac) [inline firestore-bigquery-export/scripts/gen-schema-view/src/__tests__/e2e/e2e.legacy.test.ts:81]:
```suggestion
  test("should generate a schema view based on a nestedMapSchema dataset and schema", async () => {
```
